gh-131586: Avoid refcount contention in context managers

colesbury opened the pull request:
This avoid reference count contention in the free threading build when calling special methods like `__enter__` and `__exit__`.


<!-- gh-issue-number: gh-131586 -->
* Issue: gh-131586
<!-- /gh-issue-number -->

colesbury committed:
gh-131586: Avoid refcount contention in context managers

This avoid reference count contention in the free threading build
when calling special methods like `__enter__` and `__exit__`.

colesbury commented on Python/bytecodes.c:
```diff
@@ -3345,25 +3345,29 @@ dummy_func(
-        inst(LOAD_SPECIAL, (owner -- attr, self_or_null)) {
-            assert(oparg <= SPECIAL_MAX);
-            PyObject *owner_o = PyStackRef_AsPyObjectSteal(owner);
+        op(_INSERT_NULL, (arg -- args[2])) {
```
@markshannon - this is what I was referring to in our last meeting. I had to split up `LOAD_SPECIAL` into two ops to ensure that that `frame->stackpointer` includes both operands when it's spilled before the `_PyObject_LookupSpecialMethod` call.

I'm not sure if this is a better way to do this.

colesbury commented on Tools/cases_generator/stack.py:
```diff
@@ -259,7 +259,7 @@ def pop(self, var: StackItem) -> tuple[str, Local]:
                 rename = ""
             if not popped.in_local:
-                assert popped.memory_offset is not None
+                # assert popped.memory_offset is not None
```
@markshannon - I'm not sure what to do about this.

colesbury committed:
Struggle with the cases generator

graingert committed:
Merge branch 'main' into gh-131586-lookup-special-extra

markshannon commented on Objects/typeobject.c:
```diff
@@ -2774,32 +2774,37 @@ _PyObject_LookupSpecial(PyObject *self, PyObject *attr)
+    _PyType_LookupStackRefAndVersion(Py_TYPE(self), attr, &method_and_self[0]);
+    PyObject *method_o = PyStackRef_AsPyObjectBorrow(method_and_self[0]);
+    if (method_o == NULL) {
+        return -1;
```
There is no error here, so return 0 to indicate a non-error failure.

markshannon commented on Objects/typeobject.c:
```diff
@@ -2774,32 +2774,37 @@ _PyObject_LookupSpecial(PyObject *self, PyObject *attr)
+    if (_PyType_HasFeature(Py_TYPE(method_o), Py_TPFLAGS_METHOD_DESCRIPTOR)) {
         /* Avoid temporary PyMethodObject */
-        *self_or_null = self;
+        return 0;
```
```suggestion
        return 1;
```
To indicate success.

markshannon commented on Objects/typeobject.c:
```diff
@@ -2774,32 +2774,37 @@ _PyObject_LookupSpecial(PyObject *self, PyObject *attr)
     }
-    return res;
+    PyStackRef_CLEAR(method_and_self[1]); // clear self
+    return 0;
```
```suggestion
    return 1;
```

markshannon commented on Python/bytecodes.c:
```diff
@@ -3345,25 +3345,33 @@ dummy_func(
-            if (attr_o == NULL) {
+            int err = _PyObject_LookupSpecialMethod(name, method_and_self);
+            if (err < 0) {
                 if (!_PyErr_Occurred(tstate)) {
```
```suggestion
                if (err < 0) {
```

markshannon commented on Python/bytecodes.c:
```diff
@@ -3345,25 +3345,33 @@ dummy_func(
-            PyObject *attr_o = _PyObject_LookupSpecialMethod(owner_o, name, &self_or_null_o);
-            if (attr_o == NULL) {
+            int err = _PyObject_LookupSpecialMethod(name, method_and_self);
+            if (err < 0) {
```
```suggestion
            if (err <= 0) {
```

markshannon replied:
Why not?

colesbury replied:
I was having trouble with the cases generator for a while and simplifying the control flow helped.

markshannon reviewed (commented):
Negative return values for errors should be only for errors

https://devguide.python.org/developer-workflow/c-api/#guidelines-for-expanding-changing-the-public-api

colesbury committed:
Merge branch 'main' into gh-131586-lookup-special-extra

colesbury committed:
Adjust return values

colesbury committed:
Use simpler formulation

colesbury commented on Python/generated_cases.c.h:
```diff
@@ -9268,34 +9268,29 @@
             assert(WITHIN_STACK_BOUNDS());
             _PyFrame_SetStackPointer(frame, stack_pointer);
-            PyObject *attr_o = _PyObject_LookupSpecialMethod(owner_o, name, &self_or_null_o);
+            int err = _PyObject_LookupSpecialMethod(name, method_and_self);
```
@markshannon  - I switched to the simpler formulation, but the stackpointer adjustments aren't safe here for the free threaded build. 

The `stack_pointer += -1;` means that `self` is not visible on the stack anymore when `_PyObject_LookupSpecialMethod` is called. In other words the code is generated as if the inputs are dead (okay), but the outputs are not yet alive (bad).

colesbury committed:
Split LOAD_SPECIAL into two ops again.

Make sure that `method_and_self` stays alive across the call to
_PyObject_LookupSpecialMethod. When written as a single op, the input is
marked as dead, but the outputs aren't considered alive until the end of
the op.

colesbury committed:
Merge branch 'main' into gh-131586-lookup-special-extra

colesbury committed:
Refactor error code paths

colesbury committed:
Update optimizer_bytecodes.c

colesbury committed:
Adjust formatting and ordering

brandtbucher reviewed (approved):
Yep, they look good.

mpage commented on Objects/typeobject.c:
```diff
@@ -2773,32 +2773,37 @@ _PyObject_LookupSpecial(PyObject *self, PyObject *attr)
+// is null and `method_and_self[1]` is `self`. On exit, `method_and_self[0]`
+// is the method object and `method_and_self[1]` is `self` if the method is
+// not bound.
+// Return 0 on success, -1 on error or if the method is missing.
```
This comment needs to be updated.

mpage reviewed (approved):
LGTM

markshannon reviewed (dismissed):
Try to avoid using arrays unless there is an actual array, like for `BUILD_TUPLE`.

Also, if the code generator is not doing what you think it should, file a bug report.
Either we need to document it, or fix it.

colesbury committed:
Update code comment

colesbury committed:
Merge branch 'main' into gh-131586-lookup-special-extra

colesbury committed:
Merge branch 'main' into gh-131586-lookup-special-extra
